kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -486,57 +556,171 @@ func (c *ipamController) onBlockDeleted(key model.BlockKey) {
+		}
+	}
+
+	if c.poolCache[pool.Name] == nil {
+		registerMetricVectorsForPool(pool.Name)
+		publishPoolSizeMetric(pool, poolNet)
+	}
+
+	c.poolCache[pool.Name] = pool
+}
+
+func (c *ipamController) onPoolDeleted(poolName string) {
+	// When an IP Pool is deleted, its association transitions from known to unknown.
+	for block := range c.blocksByPool[poolName] {
+		c.updatePoolForBlock(block, unknownPoolLabel)
+	}
+
+	c.unregisterMetricVectorsForPool(poolName)
+	clearPoolSizeMetric(poolName)
+
+	delete(c.blocksByPool, poolName)
+	delete(c.poolCache, poolName)
+}
+
+// Resolve the IP Pool that the Block occupies.
+func (c *ipamController) getPoolForBlock(blockCIDR string) string {
+	_, blockNet, err := net.ParseCIDR(blockCIDR)
+	if err != nil {
+		log.WithError(err).Warnf("Unable to parse block %s for pool determination", blockCIDR)
+		return unknownPoolLabel
+	}
+
+	for poolName, pool := range c.poolCache {
+		_, poolNet, err := cnet.ParseCIDR(pool.Spec.CIDR)
+		if err != nil {
+			log.WithError(err).Warnf("Failed to parse CIDR for IP Pool %s", poolName)
+			continue
+		}
+		if blockOccupiesPool(blockNet, poolNet) {
+			return poolName
+		}
+	}
+
+	return unknownPoolLabel
+}
+
+func (c *ipamController) updatePoolForBlock(blockCIDR string, newPool string) {
+	previousPool := c.poolsByBlock[blockCIDR]
+	if previousPool == newPool {
+		return
+	}
+
+	// Update pools by block
+	c.poolsByBlock[blockCIDR] = newPool
+
+	// Update blocks by pool
+	if previousPoolBlocks, ok := c.blocksByPool[previousPool]; ok {
+		delete(previousPoolBlocks, blockCIDR)
+	}
+	if c.blocksByPool[newPool] == nil {
+		c.blocksByPool[newPool] = map[string]bool{}
+	}
+	c.blocksByPool[newPool][blockCIDR] = true
+}
+
+func blockOccupiesPool(blockNet *net.IPNet, poolNet *cnet.IPNet) bool {
+	return poolNet.Covers(*blockNet)
 }
 
 func (c *ipamController) updateMetrics() {
 	log.Debug("Gathering latest IPAM state for metrics")
 
-	// Keep track of various counts so that we can report them as metrics.
-	blocksByNode := map[string]int{}
-	borrowedIPsByNode := map[string]int{}
+	// Keep track of various counts so that we can report them as metrics. These counts track legacy metrics by node.
+	legacyBlocksByNode := map[string]int{}
+	legacyBorrowedIPsByNode := map[string]int{}
 
 	// Iterate blocks to determine the correct metric values.
-	for _, kvp := range c.allBlocks {
-		b := kvp.Value.(*model.AllocationBlock)
-		if b.Affinity != nil {
-			n := strings.TrimPrefix(*b.Affinity, "host:")
-			blocksByNode[n]++
-		} else {
-			// Count blocks with no affinity as a pseudo-node.
-			blocksByNode["no_affinity"]++
-		}
-
-		// Go through each IPAM allocation, check its attributes for the node it is assigned to.
-		for _, idx := range b.Allocations {
-			if idx == nil {
-				// Not allocated.
-				continue
+	for poolName, poolBlocks := range c.blocksByPool {
+		// These counts track pool-based gauges by node for the current pool.
+		inUseAllocationsForPoolByNode := map[string]int{}
```

Comment:
this is a super super nit, but if it were me I'd shorten all of these and drop the "ForPool" from each.

e.g., `blocksByNode` - we know that we're within the context of a pool since we're within the loop, and fewer words usually means easier to read code.